vermouth/processors/molecule_grouper.py

```diff
@@ -0,0 +1,295 @@
+
+        prev_weights = flow_dict
+        cost, flow_dict = simplex_method(flow_graph)
+
+        if prev_weights == flow_dict:
+            break
+        # Step 2: make memberships and calculate new clustroid positions
+        memberships = np.zeros((len(data), num_clusters))
+        for data_idx in data_nodes:
+            flow = flow_dict[data_idx]
+            memberships[data_idx] = [flow[clust_n_idx] for clust_n_idx in clust_nodes]
+
+        # TODO: PBC
+        weights = memberships / memberships.sum(axis=0, keepdims=True)
+        clusters = weights.T.dot(data)
+
+        iter += 1
+
+    return cost/precision, clusters, memberships, iter
+
+
+def group_molecules(system, selector, size_tries=10, **kwargs):
+    """
+    Clusters molecules in `system` into groups of 4 \u00b1 1. Only molecules
+    selected with `selector` will be taken.
+
+    Parameters
+    ----------
+    system
+    selector
+    size_tries: int
+        The number of clusters to try.
+    **kwargs
+        Passed on to :func:`constrained_kmeans`.
+
+    Returns
+    -------
+    None
+        `system` is modified in place.
+    """
+    water_mols = [(mol_idx, mol) for (mol_idx, mol) in enumerate(system.molecules) if selector(mol)]
+    mol_idxs, water_mols = zip(*water_mols)
+    positions = []
+    for mol in water_mols:
+        # TODO: Select CoM vs CoG (current) from FF settings/metavars?
+        # TODO: What happens if no atom has a position?
+        position = np.average([mol.nodes[n_idx]['position']
+                               for n_idx in mol
+                               if selector_has_position(mol.nodes[n_idx])], axis=0)
+        positions.append(position)
+    positions = np.array(positions)
+    clust_size = 3  # TODO: Fetch from mapping or FF
+    num_clusters = int(np.ceil(len(water_mols)/clust_size))
+    min_clusters = max(num_clusters - size_tries//2, 0)
+    max_clusters = min(num_clusters + size_tries//2, len(positions))
+    init = kwargs.pop('init_clusters', 'fixed')
+    results = []
+    for num_clusters in range(min_clusters, max_clusters):
+        cost, clusters, memberships, niter = constrained_kmeans(
+            data=positions,
+            num_clusters=num_clusters,
+            clust_sizes=[clust_size]*num_clusters,
+            tolerances=1,
+            init_clusters=init,
+            **kwargs
+        )
+        init = np.append(clusters, [[0, 0, 0]], axis=0)
+        results.append([cost, clusters, memberships, niter])
+
+    cost, clusters, memberships, niter = min(results, key=lambda i: i[0])
+    LOGGER.info('Clustered {} water molecules into {} clusters.', len(water_mols), num_clusters)
+    counts = np.count_nonzero(memberships, axis=0)
+    LOGGER.info('Minimum/maximum number of molecules per cluster: {}/{}', counts.min(), counts.max())
+    for mol_idx in sorted(mol_idxs, reverse=True):
+        del system.molecules[mol_idx]
```

Comment:
This also assumes that all water molecules are part of a cluster?

Reply:
Every data point will always be part of exactly 1 cluster